### ISSUE TITLE
Peer listening port reports closed, but router shows port opened by UPNP and Canyouseeme.org can see serviec

### ISSUE THREAD

**Issue by dunxd** [transmission/portcheck#7]
### What is the issue?

The Peer listening port is reported as closed in Transmission settings on MacOS 13.2.1.
<img width="640" alt="image" src="https://user-images.githubusercontent.com/16608541/222229166-8915090e-3898-45e6-be9a-70dc45fd8f09.png">
However, my router shows the port as being opened by Transmission.
<img width="698" alt="image" src="https://user-images.githubusercontent.com/16608541/222229577-8f1807a9-d9f8-4e92-8568-98de87f285e1.png">
MacOS firewall is allowing incoming connections for Transmission.
<img width="439" alt="image" src="https://user-images.githubusercontent.com/16608541/222229904-89ea9da1-56b2-4a68-91f4-9df8c4dd0dd3.png">### Which application of Transmission?
Canyouseeme.org shows that the port is visible
![image](https://user-images.githubusercontent.com/16608541/222230804-18b833fb-8c3a-46d0-baef-6d3d1b5ab62a.png)

Files are transferring ok, so it may be the test in the settings panel that is not working correctly.

### Which version of Transmission?

4.0.1

**Comment by ckerr** [comment]
- What does `https://portcheck.transmissionbt.com/your-port-number-here` (with your-port-number-here replaced with the port you've opened) say?

- What version of Transmission

- What desktop UI are you using (e.g. transmission-mac, transmission-qt, transmission-gtk)

**Comment by dunxd** [comment]
Sorry - hit save before finishing draft.

I am using transmission-mac 4.0.1.

https://portcheck.transmissionbt.com/60953 shows a page only showing 0.

**Comment by dunxd** [comment]
I've used the Randomize button to generate a new port.  This is updating my router UPNP settings correctly, and canyouseeme.org can see the new port, but portcheck.transmissionbt.com/{portnumber} still shows 0.

**Comment by titer** [comment]
In case this is an IPv4 vs IPv6 problem, these two commands in a Terminal would test them separately:
```
curl -4 https://portcheck.transmissionbt.com/{port-number}; echo
curl -6 https://portcheck.transmissionbt.com/{port-number}; echo
```

**Comment by dunxd** [comment]
IPv4 version returns 1
IPv6 version returns 0

I don't believe my router supports UPNP via IPv6. If I manually open the IPv6 port on my firewall, the IPv6 command returns 1 and the Settings page shows the port is open. 

However, if I want to randomise my port each time Transmission starts, manually forwarding doesn't work well.

Torrents are still working fine, I guess over IPv4, so perhaps this is cosmetic.

**Comment by kelson42** [comment]
@titer @ckerr After our efforts to get that problem with IPv6 fixed, I finally wanted to test with my own Transmission-GTK 3.00 (bb6b5a062e) Ubuntu client and was pretty surprised and disappointed that actually it was still failing for me!
But thank you @dunxd, you have pretty much nailed it and I believe to be exactly in the same situation: 2 public IPs, one in IPv4 and one in IPv6 and `portcheck.transmissionbt.com` seems to work only with IPv4:
```
$ curl -4 https://portcheck.transmissionbt.com/9008; echo
1
$ curl -6 https://portcheck.transmissionbt.com/9008; echo
0
```
Like for @dunxd, it seems that something is wrong between Transmission and my Fritz!box as the port seems not open (or relayed) properly in IPv6 but works properly in IPV4. But I guess Transmission can work perfectly work (in IPv4), so the check should not be reported as a failure.

Maybe the portcheck.transmissionbt.com should be smarter (probably not) or then the transmission client should be. Not sure exactly what is the current logic, but I could imaging something like: test specifically IPv4 and if fail then test specifically IPv6. Not sure how exactly how behind the door the software works and if both values are relevant or not.

So at the end not sure where the ticket should stay, here or moved to Transmission software itself?

**Comment by dunxd** [comment]
If I manually set the port in my routers IPv6 to forward to my device, then `curl -6 https://portcheck.transmissionbt.com/9008; echo` does return `1`.  So the issue is that some (at least mine) routers don't support UPnP or NAT-PMP for IPv6. 

My suggestion would be that the Settings screen listening port check can handle a third outcome:

* All checks pass - green
* All checks fail - red
* One of the checks fails, the other passes - amber - ideally the text tells you if IPv4 or IPv6 failed.

**Comment by ckerr** [comment]
@dunxd this Transmission UI suggestion is still off-topic in the portcheck repo, same as it was when you suggested it a few hours ago [here](https://github.com/transmission/portcheck/pull/3#issuecomment-1465200853) :smile_cat: This is a web service that has no concept of Transmission's UI.

**Comment by titer** [comment]
AFAICT there is no ticket open about this in the transmission repo, so maybe move this one there?

If useful, the server-side component could be exposed with v4-only and v6-only hostnames to make it a little easier to test one or the other. But ultimately, I believe the problem reported here could only be addressed with changes on the application side, making it run two separate checks instead of just one

**Comment by kelson42** [comment]
@ckerr I guess your expertise will be needed to:
* Better qualify the problem: it's not clear to me how exactly the Transmission client (should) behave in such a situation 
* Make a software proposal to fix the Transmission client (at least the UI level)
* Transfer to repository `transmission/transmission` (if you agree) OR rewrite from scratch the ticket (probably better)

**Comment by dunxd** [comment]
See my comment above that was marked as off-topic. suggesting how the client UI might deal with different results from checking IPv4 and IPv6 connectivity. At the time of posting this issue, I thought the issue may be with portcheck but it seems it is more of a client UI issue so would make sense to transfer to `transmission\transmission` since I included the screenshots and discussion shows this isn't an issue with portcheck.

**Comment by beroset** [comment]
See also https://github.com/transmission/transmission/issues/5607#issuecomment-1604410611

**Comment by kelson42** [comment]
@ckerr Can we transfer this ticket to `transmission/transmission` because the portcheck works properly IMHO.

**Comment by theirongiant82** [comment]
This is still broken in 4.0.4, even if you use static port mapping without uPNP.  

Transmission will run for a while and report the port as open, but after a few days I start to see a warning from the tracker that my port is blocked.  So this is not just a UI bug, and it's more than a port check failing.  Does Transmission self-report closed ports to trackers?  Or does the tracker find out due to multiple unsuccessful attempts?

The only workaround is to quit & relaunch the app.  


Edit: I curl'ed portcheck:

```
$ curl -4 https://portcheck.transmissionbt.com/12345; echo
1
$ curl -6 https://portcheck.transmissionbt.com/12345; echo
curl: (7) Couldn't connect to server
```

My ISP does not assign IPv6 addresses.  I have an Asus AX88U router that should support IPv6 WAN addresses.

**Comment by zkrige** [comment]
Port forwarding is setup and curl -4 https://portcheck.transmissionbt.com/33335; echo shows "1"

transmission 4.0.4 just doesnt connect and downloads are extremely slow, app shows port is closed

transmission 3.0 downloads are fast and app shows "port open"

**Comment by arsradu** [comment]
Hey guys,

I have an even weirder issue here with Transmission (even with older versions like 2.9.2).
So, first of all, the port shows always as "Closed". But here's the thing...it doesn't show as client connected in the router interface, like in the initial post. It doesn't show anywhere.

UPnP is enabled, port is the default one. And if I don't use UPnP and try to set Port Forward manually, I get this error.
<img width="672" alt="Screenshot 2024-09-28 at 15 08 12" src="https://github.com/user-attachments/assets/d7bddb29-cd17-4b74-b9b8-a37ccf0892e2">

However, with another bittorrent app (qBittorrent in this case), I do get the clients showing up in the router's interface. 
<img width="733" alt="Screenshot 2024-09-28 at 14 53 49" src="https://github.com/user-attachments/assets/05f1a77f-4208-4af7-8f2a-daf9a35c6eed">

Point is, with qBittorrent, and the ports showing as open in my router, I get faaar, far better speeds, downloading the exact same thing.

So, right now, looks like an app issue to me. Looks like the app is listening to the port (I ran `lsof -i :[portNumber]` and they the port shows up as "LISTEN", but there is no "ESTABLISHED" connetion), so, I think it might not be connecting through it).

I also want to mention that I tested:
`curl -4 https://portcheck.transmissionbt.com/PortNumber; echo`
and
`curl -6 https://portcheck.transmissionbt.com/PortNumber; echo`
and they both return `0`.

CanYouSeeMe.org shows: _Error: I could not see your service on [IP] on port (51413) Reason: Connection timed out_

But again, since it works fine with other apps, I think the problem is likely somewhere else.

I feel like this might be a combination of app issue and network setup.
For reference, my setup looks like this: ISP > ONT > My Router (tp-link C80 in case it makes a difference) > PC/laptop etc.

I tested multiple versions, and this also occurs with previous OSes as well. So, I think it's probably just an unusual setup that hasn't been taken into account yet.

I also tested the exact same app on another WiFi network, with a different setup (only ONT, which also doubles as wifi router) and the issue did NOT occur. Port showed up as Open, and everything ok.

I also read about other people's experiences and tested the theory of the IPv6 being the cause of this. However, I have IPv6 already disabled in my router. And even with it enabled, it doesn't make any difference to the issue.

So yeah....I know probably most people won't have this issue, if they're just using the default ONT, and it could also be ISP dependent. Mine is using PPPoE, just as a reference, the other one was not (not sure what exactly is it using, but it's not PPPoE).

Anyway, I wrote a lot, and I apologise for this. But I've had this issue for a looong time and never had the time to test different scenarios until recently. So I hope that helps.

If you have any questions or things you want me to try, please, let me know. I'd be glad to help, however I can.

**Comment by dunxd** [comment]
The problem as far as I can see it is that the UI shows a problem if IPv6 is enabled on the client, but for some reason cannot connect (i.e. the router blocks it) _even if IPv4 is working perfectly_. This makes troubleshooting really hard. 

In my opinion the UI should show seperate indicators for IPv4 and IPv6 connections, so it is possible for users to see what is not working and take appropriate action. Many users won't care of IPv6 is working or not, if IPv4 is, so this false indication of a global problem is wasting people's time.

You might want to completely disable IPv6 on your client as well as on your router, since I think if it is enabled on the client then Transmission always shows a problem.

**Comment by theirongiant82** [comment]
Except it's not a false indicator.  Ports appear open for about 24 hours after restarting Transmission.  During that time, Transmission reports the port is open.

If a torrent is added after 24 hours or so, the **_tracker_** reports that it cannot reach my client and to try unblocking ports.  Transmission now shows the port as closed. 

They're already open from the firewall.  I have to assume that Transmission sends its open ports in the announce.

**Comment by dunxd** [comment]
I'm not saying it will fix your problem, but if you don't need it *turn IPv6 off on your computer*. That's one less thing to worry about and it's only when you do this that the port indicator in Transmission is helpful for troubleshooting. 

Have you tried the other methods mentioned above to manually troubleshoot?

**Comment by theirongiant82** [comment]
Disabling IPv6 across a LAN isn't practical in homes with Apple HomeKit, Google Home, and other popular IoT platforms that use IPv6 for broadcast and control.  Even HomePods and other smart speakers used for audio playback will utilize IPv6 on LAN.  So does the AppleTV, a popular accessory for those watching movies and TV shows.  AirPlay and Mirroring rely on IPv6 to establish the session.

I would assume that at least 20% of your Mac users have something in their house that depends on IPv6, and disabling that stack just to fix Transmission would be unacceptable.

**Comment by dunxd** [comment]
You seem to be mistaking me for the developer of Transmission. I am a user like you. 

I suggested a step you could take to _troubleshoot_ the problem you added to the Issue I created last year. 

I'm *not* suggesting that _every_ user of Transmission should turn off IPv6 on their client. I am suggesting that _you_ turn it off _on the computer you run Transmission on_ to see if that is any way related to the problem you are having. You can of course suit yourself, but in that case I suggest you create your own issue instead of hijacking this one.

**Comment by tim88435** [comment]
Using transmission-gtk 4.0.5-1build5mint1+wilma on Mint 22.2 Cinnamon 6.4.8, issue still persists
Network settings in the app says port closed, but portcheck says open on ipv4, closed on ipv6
I've tried manual and uPNP, no difference.

**Comment by titer** [comment]
This is already tracked in https://github.com/transmission/transmission/issues/5607 for the application side of the issue. The portcheck service itself handles both IPv4 and IPv6 requests fine.
